src/conventional/sonde_bufr2ioda.py

```diff
@@ -0,0 +1,639 @@
+    if data.dtype == np.float64:
+        return all(x == double_missing_value for x in data)
+    elif data.dtype == np.int32:
+        return all(x == int_missing_value for x in data)
+    elif data.dtype == np.int64:
+        return all(x == long_missing_value for x in data)
+    elif data.dtype.kind in {'U', 'S'}:
+        return all(x == string_missing_value for x in data)
+
+    logging.warning("data type not determined, returning False")
+    return False
+
+
+def get_normalized_bit(value, bit_index):
+    return (value >> bit_index) & 1
+
+
+def def_significance_table():
+
+    significance_table = {
+        "pressure level originally indicated by height as the veritcal coordinate": 1,
+        "freezing level": 2,
+        "level determined by regional decision": 3,
+        "top of wind sounding": 4,
+        "end of missing wind data": 5,
+        "beginning of missing wind data": 6,
+        "end of missing humidity data": 7,
+        "beginning of missing humidity data": 8,
+        "end of missing temperature data": 9,
+        "beginning of missing temperature data": 10,
+        "significant wind level": 11,
+        "significant humidity level": 12,
+        "significant temperature level": 13,
+        "maximum wind level": 14,
+        "tropopause level": 15,
+        "standard level": 16,
+        "surface": 17,
+    }
+
+    return significance_table
+
+
+def specialty_time (bufr, tvals):
+
+    bad_date = False
+    vals = []
+    result = []
+
+    try:
+        vals = ecc.codes_get_array(bufr, 'year')
+        year = vals[0]
+    except ecc.KeyValueNotFoundError:
+        logging.warning("Caution, no data for year")
+        bad_date = True
+
+    try:
+        vals = ecc.codes_get_array(bufr, 'month')
+        month = vals[0]
+    except ecc.KeyValueNotFoundError:
+        logging.warning("Caution, no data for month")
+        bad_date = True
+
+    try:
+        vals = ecc.codes_get_array(bufr, 'day')
+        day = vals[0]
+    except ecc.KeyValueNotFoundError:
+        logging.warning("Caution, no data for day")
+        bad_date = True
+
+    try:
+        vals = ecc.codes_get_array(bufr, 'hour')
+        hour = vals[0]
+    except ecc.KeyValueNotFoundError:
+        logging.warning("Caution, no data for hour")
+        bad_date = True
+
+    try:
+        vals = ecc.codes_get_array(bufr, 'minute')
+        minute = vals[0]
+    except ecc.KeyValueNotFoundError:
+        logging.warning("Caution, no data for minute")
+        bad_date = True
+
+    second = 0
+    try:
+        vals = ecc.codes_get_array(bufr, 'second')
+        second = vals[0]
+        if (second < 0 or second > 59):
+            second = 0
+    except ecc.KeyValueNotFoundError:
+        logging.info("Caution, no data for second")
+
+    if (year < 1900 or year>2499 or month < 1 or month > 12 or day < 1 or day > 31 or hour < 0 or hour > 23 or minute < 0 or minute > 59):
+        bad_date = True
+
+    if bad_date:
```

Comment:
Sorry if I'm not getting the logic right in this section of the code.

It looks like, when bad_data is False, time_offset is set to the number of seconds from the epoch to the first datetime value from the bufr file. Then all subsequent datetime values from the bufr file are adjusted to offsets from the epoch (via the time_offset value). This part looks good.

It seems a bit risky to silently "repair" bad datetime values (ie out of the range 0 to 6 hours). Would it be better to put in missing values?

When bad_data is False, it looks like to me all of the datetimes from the bufr file are getting translated to being offsets from Jan 1, 1900 0Z instead of the more current time you would likely have if the first datetime was valid.

Was this done to force the ioda reader to throw away all of the obs? If so, would it be better to just remove this bufr message from the output file altogether? Perhaps you could check the first datetime value at the start of the read_bufr_message function and skip the processing of the message if it is invalid.

Reply:
The point you're making is valid.  Here is the problem:  there are actually no quality checks at all in the creation of BUFR messages.  Quite sad but true.  So one sounding I found has a year of "22" instead of "2022."  It's actually ridiculous in this modern age how poorly quality controlled the *inputting* of data is in the first place.  Another example that is possible is specifying the 31st of February - there is almost nothing to prevent this insanity.  The logic behind my choice of setting a default bad date of 1900-01-01T00:00:00Z is simply that we do not have any such data back then yet the datetime object in Python will permit the calculation of time offset from epoch of that date.  Once in UFO, a year of 1900 will be rejected anyway.

My attempt here is to try and prevent the ioda-converter from failing given as many possible pathological inputs as possible - yet I do know that if someone codes a BUFR message with month=2 and day=31, the python datetime object will choke on it since I am not using anything like try-except.

Reply:
Thanks for the feedback @gthompsnJCSDA. I think what you are saying makes sense and sounds reasonable to me. I like the aspect of deferring the choice to "rescue" the data to UFO. Thanks for your patience with this.